// @sudoku/stores/domain.js
import { writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';

// ============================
// 常量与默认状态
// ============================

const EMPTY_GRID = Array(9).fill(null).map(() => Array(9).fill(0));
const DEFAULT_EXPLORE_STATUS = {
    active: false,
    knownFailed: false,
    hasConflict: false,
    failedReason: null,
    failedCount: 0,
};

// ============================
// 初始化快照
// ============================

function makeInitialState() {
    return {
        puzzle: EMPTY_GRID,
        grid: EMPTY_GRID,
        canUndo: false,
        canRedo: false,
        mode: 'normal',
        isExploring: false,
        exploreStatus: { ...DEFAULT_EXPLORE_STATUS },
        lastHint: null,
    };
}

// ============================
// Svelte store 封装
// ============================

// createGameStore 对领域层的 Game 对象进行封装，并暴露一个兼容 Svelte 的
// 响应式 store。每次状态变更都会调用 _sync()，用新的普通对象快照替换 store 的值，从而触发 Svelte 的变更检测。
function createGameStore() {
    // 内部领域对象实例（非响应式），所有业务能力都在它上面执行。
    let _game = null;
    // 最近一次提示（用于 UI 展示提示原因与等级）。
    let _lastHint = null;

    // 对外暴露给 UI 的响应式快照。
    // 只放“可序列化的普通数据”，避免把复杂对象（例如 class 实例）直接放入 Svelte store。
    const _internal = writable(makeInitialState());

    // ============================
    // 同步与初始化
    // ============================

    /**
     * 把当前 _game 状态同步到响应式 store。
     *
     * 设计要点：
     * - 永远 set 一个新对象，确保触发 Svelte 订阅更新
     * - 如果领域层未实现某能力（兼容旧版本），回退到默认值
     */
    function _sync() {
        if (!_game) return;

        const exploreStatus = _game.getExploreStatus();

        // 通过重新赋值（而不是原地修改）来通知 Svelte store 状态已变化
        _internal.set({
            puzzle: _game.getSudoku().getPuzzle(),
            grid: _game.getSudoku().getGrid(),
            canUndo: _game.canUndo(),
            canRedo: _game.canRedo(),
            mode: _game.getMode(),
            isExploring: _game.isExploring(),
            exploreStatus,
            lastHint: _lastHint,
        });
    }

    function start(puzzleGrid) {
        _game = createGame({ sudoku: createSudoku(puzzleGrid) });
        _sync();
    }

    // ============================
    // 基础操作
    // ============================

    function guess(move) {
        if (!_game) return;
        _game.guess(move);
        _sync();
    }

    function getCellCandidates(row, col) {
        if (!_game) return [];
        return _game.getCellCandidates(row, col);
    }

    // ============================
    // 提示相关
    // ============================

    /**
     * 查询“候选提示”（只读，不修改棋盘）。
     *
     * @param {object} options 提示选项（例如 { row, col }）
     * @returns {object|null} 提示对象或 null
     */
    function getCandidatesHint(options = {}) {
        if (!_game) return null;
        const hint = _game.getCandidatesHint(options);
        _lastHint = hint || null;
        _sync();
        return hint;
    }

    /**
     * 查询“下一步提示”信息（只读，不修改棋盘）。
     *
     * @returns {object|null} 提示对象或 null
     */
    function getNextValueHint() {
        if (!_game) return null;
        const hint = _game.getNextValueHint();
        _lastHint = hint || null;
        _sync();
        return hint;
    }

    /**
     * 应用下一步提示到棋盘。
     *
     * 流程：
    * 1) 向领域层请求提示（尽量拿可填写值）
     * 2) 校验 row/col/value 是否完整
     * 3) 调用 _game.guess() 写入并 _sync() 刷新 UI
     *
     * @returns {object|null} 提示对象；无可用提示时返回 null
     */
    function applyNextValueHint() {
        if (!_game) return null;
        const hint = _game.getNextValueHint();
        if (!hint) return null;
        const value = Number.isInteger(hint.directValue) ? hint.directValue : hint.value;
        if (Number.isInteger(hint.row) && Number.isInteger(hint.col) && Number.isInteger(value)) {
            _game.guess({ row: hint.row, col: hint.col, value });
        }
        _lastHint = hint;
        _sync();
        return hint;
    }

    // ============================
    // Undo / Redo
    // ============================

    /** 撤销一步（当前模式：normal 或 explore），随后同步 UI。 */
    function undo() {
        if (!_game) return;
        _game.undo();
        _sync();
    }

    /** 重做一步（当前模式：normal 或 explore），随后同步 UI。 */
    function redo() {
        if (!_game) return;
        _game.redo();
        _sync();
    }

    // ============================
    // Explore 模式控制
    // ============================

    /**
     * 进入 Explore 模式。
     *
     * @returns {boolean} 是否成功进入（已在探索中时一般返回 false）
     */
    function startExplore() {
        if (!_game) return false;
        const ok = _game.startExplore();
        _sync();
        return ok;
    }

    /**
     * 提交 Explore 结果到主局面。
     *
     * @returns {boolean} 是否提交成功
     */
    function commitExplore() {
        if (!_game) return false;
        const ok = _game.commitExplore();
        _sync();
        return ok;
    }

    /**
     * 放弃 Explore 并回滚到探索起点。
     *
     * @returns {boolean} 是否回滚成功
     */
    function abandonExplore() {
        if (!_game) return false;
        const ok = _game.abandonExplore();
        _sync();
        return ok;
    }

    /**
     * 读取 Explore 状态快照。
     *
     * 在 _game 未初始化时返回默认状态，保证 UI 可安全渲染。
     */
    function getExploreStatus() {
        if (!_game) return { ...DEFAULT_EXPLORE_STATUS };
        return _game.getExploreStatus();
    }

    /**
     * 手动标记当前 Explore 局面为失败（可用于交互按钮或调试）。
     *
     * @param {string} reason 失败原因标签
     * @returns {boolean} 是否标记成功
     */
    function markExploreFailed(reason = 'manual') {
        if (!_game) return false;
        const ok = _game.markExploreFailed(reason);
        _sync();
        return ok;
    }

    // ============================
    // 对外 API
    // ============================

    /**
     * 对外 API：
     * - subscribe：给 Svelte 组件订阅状态
     * - 其余方法：把 UI 事件桥接到领域层 Game 实例
     */
    return {
        subscribe: _internal.subscribe,
        start,
        guess,
        getCellCandidates,
        getCandidatesHint,
        getNextValueHint,
        applyNextValueHint,
        undo,
        redo,
        startExplore,
        commitExplore,
        abandonExplore,
        getExploreStatus,
        markExploreFailed,
    };
}

export const gameStore = createGameStore();
