// @sudoku/stores/domain.js
import { writable, derived } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';

const EMPTY_GRID = Array(9).fill(null).map(() => Array(9).fill(0));

// createGameStore 对领域层的 Game 对象进行封装，并暴露一个兼容 Svelte 的
// 响应式 store。每次状态变更都会调用 _sync()，用新的普通对象快照替换store 的值，从而触发 Svelte 的变更检测。
function createGameStore() {
    let _game = null;

    const _internal = writable({
        puzzle: EMPTY_GRID,
        grid: EMPTY_GRID,
        canUndo: false,
        canRedo: false,
    });

    function _sync() {
        if (!_game) return;
        // 通过重新赋值（而不是原地修改）来通知 Svelte store 状态已变化
        _internal.set({
            puzzle: _game.getSudoku().getPuzzle(),
            grid: _game.getSudoku().getGrid(),
            canUndo: _game.canUndo(),
            canRedo: _game.canRedo(),
        });
    }

    return {
        subscribe: _internal.subscribe,

        // 使用给定题面网格（9x9 数字数组）开始新游戏
        start(puzzleGrid) {
            _game = createGame({ sudoku: createSudoku(puzzleGrid) });
            _sync();
        },

        // 记录玩家落子并同步到 UI
        guess(move) {
            if (_game) { _game.guess(move); _sync(); }
        },

        // 撤销最近一次玩家操作
        undo() {
            if (_game) { _game.undo(); _sync(); }
        },

        // 重做最近一次被撤销的操作
        redo() {
            if (_game) { _game.redo(); _sync(); }
        },
    };
}

export const gameStore = createGameStore();
